### PR TITLE
Clean up edge cases in `isinstance`/`issubclass`.

### PR DESCRIPTION
**Pull request by DimitrisJim**

### CODE REVIEW

**Comment by DimitrisJim** [comment]
@Snowapril wanna review these changes?

File: vm/src/vm.rs [hunk 1]
```diff
@@ -944,14 +944,50 @@ impl VirtualMachine {
...
+            // Only mask AttributeErrors.
+            if e.class().is(&self.ctx.exceptions.attribute_error) {
+                self.new_type_error(msg)
+            } else {
+                e
+            }
+        })
+    }
+
     pub fn _isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
```
Comment (Snowapril) [inline vm/src/vm.rs:960]:
```suggestion
    fn recursive_isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
```
Totally looks good to me 😊 How about rename `_isinstance` to `recursive_isinstance`? As cpython and `issubclass` use `recursive_*` naming for that case, I think it is a little more reasonable.

File: vm/src/vm.rs [hunk 2]
```diff
@@ -944,14 +944,50 @@ impl VirtualMachine {
         }
     }
 
+    // Equivalent to check_class. Masks Attribute errors (into TypeErrors) and lets everything
+    // else go through.
+    fn _check_cls(&self, cls: &PyObjectRef, msg: String) -> PyResult {
```
Comment (youknowone) [inline vm/src/vm.rs:949]: msg part need to be a closure to avoid unnecessary formatting

```suggestion
    fn _check_cls<F>(&self, cls: &PyObjectRef, msg: F) -> PyResult
    where
        F: Fn() -> String
    {
```

File: vm/src/vm.rs [hunk 3]
```diff
@@ -944,14 +944,50 @@ impl VirtualMachine {
...
+        })
+    }
+
     pub fn _isinstance(&self, obj: &PyObjectRef, cls: &PyObjectRef) -> PyResult<bool> {
         if let Ok(typ) = PyTypeRef::try_from_object(self, cls.clone()) {
-            Ok(obj.isinstance(&typ))
+            if obj.class().issubclass(typ.clone()) {
+                Ok(true)
+            } else if let Ok(icls) =
+                PyTypeRef::try_from_object(self, self.get_attribute(obj.clone(), "__class__")?)
```
Comment (youknowone) [inline vm/src/vm.rs:965]: is this line meaning `obj.class()` sometimes different from `obj.__class__`?

File: vm/src/vm.rs (hunk 1 above)
Comment (DimitrisJim) [reply vm/src/vm.rs:960]: Good idea!

File: vm/src/vm.rs (hunk 3 above)
Comment (DimitrisJim) [reply vm/src/vm.rs:965]: Yup, check out the test cases for examples where this is manipulated.

File: vm/src/vm.rs (hunk 2 above)
Comment (DimitrisJim) [reply vm/src/vm.rs]: This generally follows the pattern found in the `vm.new_exception_type(&self, msg: String)`. Is there any benefit in passing a closure that calls the `format!` macro itself?

Comment (youknowone) [reply vm/src/vm.rs]: Yes, the error message is only used when conditionally error raised. But if we format the argument itself, the format is always executed regardless the condition it is required or not.

File: vm/src/vm.rs (hunk 3 above)
Comment (youknowone) [reply vm/src/vm.rs:965]: when `__class__` is replaced by python code, I think `obj.class()` need to be replaced rather than checked by here. I am good with this way for now, but could you add a FIXME comment here?

File: vm/src/vm.rs (hunk 2 above)
Comment (DimitrisJim) [reply vm/src/vm.rs]: Ah true, address that.

File: vm/src/vm.rs (hunk 3 above)
Comment (DimitrisJim) [reply vm/src/vm.rs:968]: Oh, sorry, I just saw the comment. Are you sure that's possible? I don't think `CPython` allows Python code modifying `tp_type` (and `obj.class()` grabs the equivalent of that if I'm not mistaken). It's the reason why on the next line we check:

```
if icls.is(&obj.class())
```

this way, if `__class__` is something different, we check if that is a subclass of `typ`.

**Review by Snowapril — APPROVED** [review]

File: vm/src/vm.rs (hunk 3 above)
Comment (youknowone) [reply vm/src/vm.rs:968]: Thanks, I think that's possible but honestly I didn't checked how CPython handle it.

**Review by youknowone — APPROVED** [review]
